pkg/provider/pulumi/pulumi.go

```diff
@@ -16,49 +16,115 @@ package pulumi
-	x, _, err := c.escClient.OpenEnvironment(ctx, c.organization, c.environment, "", 5*time.Minute)
+func (c *client) GetSecret(_ context.Context, ref esv1beta1.ExternalSecretDataRemoteRef) ([]byte, error) {
+	env, err := c.escClient.OpenEnvironment(c.authCtx, c.organization, c.environment)
 	if err != nil {
 		return nil, err
 	}
-	value, err := c.escClient.GetOpenProperty(ctx, c.organization, c.environment, x, ref.Key)
+	value, _, err := c.escClient.ReadEnvironmentProperty(c.authCtx, c.organization, c.environment, env.GetId(), ref.Key)
 	if err != nil {
 		return nil, err
 	}
-	return utils.GetByteValue(value.ToJSON(false))
+	return utils.GetByteValue(value.GetValue())
 }
 
-func (c *client) PushSecret(_ context.Context, _ *corev1.Secret, _ esv1beta1.PushSecretData) error {
-	return errors.New(errPushSecretsNotSupported)
+func (c *client) PushSecret(_ context.Context, secret *corev1.Secret, data esv1beta1.PushSecretData) error {
+	value := secret.Data[data.GetSecretKey()]
+
+	updatePayload := &esc.EnvironmentDefinition{
+		Values: &esc.EnvironmentDefinitionValues{
+			AdditionalProperties: map[string]interface{}{
+				data.GetRemoteKey(): string(value),
+			},
+		},
+	}
+	_, oldValues, err := c.escClient.OpenAndReadEnvironment(c.authCtx, c.organization, c.environment)
+	if err != nil {
+		return fmt.Errorf(errReadEnvironment, err)
+	}
+	updatePayload.Values.AdditionalProperties = mergeMaps(oldValues, updatePayload.Values.AdditionalProperties)
+	_, err = c.escClient.UpdateEnvironment(c.authCtx, c.organization, c.environment, updatePayload)
+	if err != nil {
+		return fmt.Errorf(errPushSecrets, err)
+	}
+
+	return nil
+}
+
+func mergeMaps(map1, map2 map[string]interface{}) map[string]interface{} {
+	mergedMap := make(map[string]interface{})
+
+	// Helper function to merge nested maps
+	var mergeNestedMap func(m map[string]interface{}, keys []string, value interface{})
+	mergeNestedMap = func(m map[string]interface{}, keys []string, value interface{}) {
+		if len(keys) == 1 {
+			m[keys[0]] = value
+			return
+		}
+		key := keys[0]
+		if _, exists := m[key]; !exists {
+			m[key] = make(map[string]interface{})
+		} else {
+			if _, ok := m[key].(map[string]interface{}); !ok {
+				m[key] = make(map[string]interface{})
+			}
+		}
+		nestedMap := m[key].(map[string]interface{})
+		mergeNestedMap(nestedMap, keys[1:], value)
+	}
+
+	// Add all key-value pairs from the first map to the merged map
+	for key, value := range map1 {
+		if strings.Contains(key, ".") {
+			parts := strings.Split(key, ".")
+			mergeNestedMap(mergedMap, parts, value)
+		} else {
+			mergedMap[key] = value
+		}
+	}
+
+	// Add all key-value pairs from the second map to the merged map,
+	// overwriting values for any existing keys
+	for key, value := range map2 {
+		if strings.Contains(key, ".") {
+			parts := strings.Split(key, ".")
+			mergeNestedMap(mergedMap, parts, value)
+		} else {
+			mergedMap[key] = value
+		}
+	}
```

Comment:
This also could be simplified with a helper function and passing in the two maps.

Reply:
reworked this code as it is to fragile!